theurgy.lic

```diff
@@ -1,186 +1,449 @@
-=end
+custom_require.call(%w[common common-arcana common-items common-money common-travel drinfomon spellmonitor equipmanager])
 
-custom_require.call(%w[common common-items common-money common-travel drinfomon spellmonitor equipmanager])
-
-class Theurgy
+class TheurgyActions
   include DRC
   include DRCI
-  include DRCM
   include DRCT
+  include DRCM
+  include DRCA
+  
+  def initialize(settings, data)
+    @settings = settings
+    @hometown = @settings.hometown
+    @data = data
+
+    @theurgy_supply_container = @settings.theurgy_supply_container
+    @water_holder = @settings.water_holder
+    @immortal_aspect = @settings.immortal_aspect
+    @flint_lighter = @settings.flint_lighter
+    @safe_room = @settings.safe_room
+
+    @rituals = get_rituals
+    @communes = get_communes
+
+    annotate_item_presence(@rituals + @communes)
+    
+    @rituals.reject! do |r|
+      r[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+    @communes.reject! do |c|
+      c[:items].any? { |i| !i[:on_hand] && !i[:shop] }
+    end
+  end
 
-  ######################## GET/PUT ########################
-
-  def put_in_container(item)
-    bput "put #{item} in my #{@theurgy_supply_container}", 'You put', 'What were you referring to'
+  def get_rituals
+    rituals = all_rituals
+    blacklist = @settings.theurgy_blacklist || []
+    rituals.reject! { |r| blacklist.include? r[:method].to_s }
+    whitelist = @settings.theurgy_whitelist || []
+    unless whitelist.empty?
+      rituals.select! { |r| whitelist.include? r[:method].to_s }
+    end
+    rituals
+  end
+  
+  def all_rituals
+    rituals = []
+    if @data['almsbox'] && @settings.tithe
+      rituals << {
+        method: :tithe,
+      }
+    end
+    if @hometown == 'Crossing'
+      if DRStats.circle >= 30
+        rituals << {
+          method: :study_wall,
+        }
+      end
+      rituals << {
+        method: :refectory,
+      }
+    end
+    if DRSkill.getrank('Mechanical Lore') > 140 && @immortal_aspect && @data['limb_foraging']
+      # TODO: allow other knife nouns 
+      rituals << {
+        method: :carve_bead,
+        items: [
+          { name: 'carving knife', any_container: true },
+          { name: 'shaper', any_container: true},
+          { name: 'prayer chain', any_container: true },
+        ],
+      }
+    end
+    if @data['gather_sirese'] && @data['plant_sirese']
+      rituals << {
+        method: :sirese_seed,
+      }
+    end
+    if @data['bath']
+      herbs = @data['herbs'] || ['sage', 'lavender']
+      rituals << {
+        method: :bathe,
+        items: herbs.map { |h| { name: h, shop: @data['theurgy_supplies'] } },
+      }
+    end
+    if @immortal_aspect
+      ritals << {
+        method: :meditate_bead,
+        items: [
+          { name: 'prayer chain', any_container: true },
+        ],
+      }
+    end
+    rituals << {
+      method: :pray_badge,
+    }
+    # TODO: prayer mat support
```

Comment:
When will this be implemented?

Reply:
When I get a prayer mat 😃 

But actually, I'm planning to harvest the code from "Theurgy 2.0" that covers this as my next PR.

Reply:
Why not just include it in this PR?

Reply:
If the old script didn't support prayer mats, I wouldn't worry about getting it in as part of this PR

Reply:
I don't understand that thinking. Theurgy already exists and works. This is an update to add new features and make it more robust, how would adding prayer mat support be outside of the scope?

Reply:
> Replaces the Crossing-specific theurgy.lic with a region-independent script. Reimplements all functionality of the original including support for bead carving and magical research.

I interpreted the PR description to mean that the goal is feature parity with the old script, but in a region-independent fashion.

Reply:
I just read the title. Still think we should implement new things as we see them.

Reply:
Also, this is more than adding region-independent code, this is an entire rewrite of the logic. I think the scope of this PR is larger than the basic addition of new areas.

Reply:
The goal is to reimplement but make more robust and region independent.  I want to add all of these other features as well, but in a separate PR.  I think they're all good ideas.

Part of why I'm hesitant to include them here is that I can't test them myself.

Reply:
That's fine, I just didn't understand the scope of the PR.